python-avd/pyavd/_eos_designs/structured_config/network_services/router_internet_exit.py

```diff
@@ -19,28 +21,200 @@ class RouterInternetExitMixin(Protocol):
-            #       This works for zscaler but later we may need to use some sorting intelligence as order matters.
-            for connection in connections:
-                exit_group_name = connection["exit_group"]
-                exit_groups.obtain(exit_group_name).local_connections.append_new(name=connection["name"])
-                # Recording the exit_group in the policy
-                policies.obtain(policy.name).exit_groups.append_new(name=exit_group_name)
-
-            if policy.fallback_to_system_default:
-                policies.obtain(policy.name).exit_groups.append_new(name="system-default-exit-group")
-
-        self.structured_config.router_internet_exit._update(policies=policies, exit_groups=exit_groups)
+        if not (internet_exit_policy_name := input_topology.internet_exit.policy):
+            return
+
+        if internet_exit_policy_name in self.structured_config.router_internet_exit.policies:
+            # We have already added this policy
+            return
+
+        if internet_exit_policy_name not in self.inputs.cv_pathfinder_internet_exit_policies:
+            msg = (
+                f"The internet exit policy {internet_exit_policy_name} configured under "
+                f"`wan_virtual_topologies.policies[name={policy_name}].internet_exit.policy` "
+                "is not defined under `cv_pathfinder_internet_exit_policies`."
+            )
+            raise AristaAvdInvalidInputsError(msg)
+
+        internet_exit_policy = self.inputs.cv_pathfinder_internet_exit_policies[internet_exit_policy_name]
+
+        # duplicate with some function in utils_wan.
+        # TODO: Add support for port_channels
+        local_wan_l3_interfaces = EosDesigns._DynamicKeys.DynamicNodeTypesItem.NodeTypes.NodesItem.L3Interfaces(
+            [
+                wan_interface
+                for wan_interface in self.shared_utils.wan_interfaces
+                if internet_exit_policy.name in wan_interface.cv_pathfinder_internet_exit.policies
+            ]
+        )
+        if not local_wan_l3_interfaces:
+            # No local interface for this policy
+            # implies policy present in input yml, but not associated with any interface yet
+            # TODO: Decide if we should raise here instead
+            return
+
+        # enabling monitor connectivity
+        self.structured_config.monitor_connectivity.shutdown = False
+        # enabling router_service_insertion once
+        self.structured_config.router_service_insertion.enabled = True
+
+        if internet_exit_policy.type == "direct":
+            self._set_direct_internet_exit_policy(internet_exit_policy, local_wan_l3_interfaces)
+        else:  # internet_exit_policy.type == "zscaler"
+            self._set_zscaler_internet_exit_policy(internet_exit_policy, local_wan_l3_interfaces)
+            # TODO: Technically we need this ONLY if there is ANY Zscaler policy
```

Comment:
What is the TODO here? You are already nesting it under the zscaler branch.

Reply:
yeah but it will be called for every zscaler routing policy we find - for now the way it is organized I would need to bubble this up to take a decision later in the main contributor but that means bubbling up boolean quite through a few hoops so I though taking the hit here was ok.

Reply:
ah ok, or call a helper directly from the nested function